third_party/terraform/utils/retry_utils.go

```diff
@@ -0,0 +1,85 @@
+package google
+
+import (
+	"log"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"google.golang.org/api/googleapi"
+)
+
+func retry(retryFunc func() error) error {
+	return retryTime(retryFunc, 1)
+}
+
+func retryTime(retryFunc func() error, minutes int) error {
+	return retryTimeDuration(retryFunc, time.Duration(minutes)*time.Minute)
+}
+
+func retryTimeDuration(retryFunc func() error, duration time.Duration, errorRetryPredicates ...func(e error) (bool, string)) error {
+	return resource.Retry(duration, func() *resource.RetryError {
+		err := retryFunc()
+		if err == nil {
+			return nil
+		}
+		for _, e := range getAllTypes(err, &googleapi.Error{}, &url.Error{}) {
+			if isRetryableError(e, errorRetryPredicates) {
+				return resource.RetryableError(e)
+			}
+		}
+		return resource.NonRetryableError(err)
+	})
+}
+
+func getAllTypes(err error, args ...interface{}) []error {
+	var result []error
+	for _, v := range args {
+		subResult := errwrap.GetAllType(err, v)
+		if subResult != nil {
+			result = append(result, subResult...)
+		}
+	}
+	return result
+}
+
+func isRetryableError(err error, retryPredicates []func(e error) (bool, string)) bool {
+	// These operations are always hitting googleapis.com - they should rarely
+	// time out, and if they do, that timeout is retryable.
+	if urlerr, ok := err.(*url.Error); ok && urlerr.Timeout() {
+		log.Printf("[DEBUG] Dismissed an error as retryable based on googleapis.com target: %s", err)
+		return true
+	}
+
+	if gerr, ok := err.(*googleapi.Error); ok {
```

Comment:
It could be nice to move the base retry cases to predicates as a followup and use them as defaults for consistency with other retry predicates.

Reply:
yup, sg!